File: android/src/main/kotlin/com/oguzhnatly/flutter_carplay/FCPProtocols.kt
package com.oguzhnatly.flutter_carplay

import androidx.car.app.CarContext

/** Protocol representing a generic template in the Flutter CarPlay (FCP) framework. */
abstract class FCPTemplate {
    /// The unique identifier for the template.
    lateinit var elementId: String

    /// The unique identifier for the template's back button.
    var backButtonElementId: String? = null

    /// Update the template on UI.
    var onInvalidate: (() -> Unit) = {}

    /// Returns the underlying CPTemplate instance configured with the specified properties.
    abstract fun getTemplate(): CPTemplate
}

/** Protocol representing a root template in the Flutter CarPlay (FCP) framework. */
abstract class FCPRootTemplate : FCPTemplate()

/** Protocol representing a template that can be presented in the Flutter CarPlay (FCP) framework. */
abstract class FCPPresentTemplate : FCPTemplate()

/**
 * Converts the FCPTemplate to a FCPScreen.
 *
 * @param carContext the CarContext used to create the FCPScreen
 * @return the FCPScreen created from the FCPTemplate
 */
fun FCPTemplate.toScreen(carContext: CarContext) = FCPScreen(carContext, this)

File: ios/Classes/FCPProtocols.swift
//
//  FCPProtocols.swift
//  flutter_carplay
//
//  Created by Oğuzhan Atalay on 25.08.2021.
//
import CarPlay

public protocol FCPPresentTemplate: AnyObject {}

public protocol FCPRootTemplate: AnyObject {
  var get: CPTemplate { get }
  var elementId: String { get }
}
